src/VectorBundle.jl

```diff
@@ -0,0 +1,221 @@
+
+"""
+    VectorSpaceType
+
+Abstract type for tangent space, cotangent space, their tensor products,
+exterior products etc.
+"""
+abstract type VectorSpaceType end
+
+struct TangentSpaceType <: VectorSpaceType end
+struct CotangentSpaceType <: VectorSpaceType end
+
+TCoTSpaceType = Union{TangentSpaceType, CotangentSpaceType}
+
+const TangentSpace = TangentSpaceType()
+const CotangentSpace = CotangentSpaceType()
+
+"""
+    similar_result_type(M::Manifold, VS::VectorSpaceType, f, args::NTuple{N,Any}) where N
+
+Returns type of element of the array that will represent the result of
+function `f` for representing an operation with result in the vector space `VS`
+for manifold `M` on given arguments (passed at a tuple).
+"""
+function similar_result_type(M::Manifold, VS::VectorSpaceType, f, args::NTuple{N,Any}) where N
+    T = typeof(reduce(+, one(eltype(eti)) for eti ∈ args))
+    return T
+end
+
+"""
+    similar_result(M::Manifold, VS::VectorSpaceType, f, x...)
+
+Allocates an array for the result of function `f` that is
+an element of the vector space of type `VS` on manifold `M`
+and arguments `x...` for implementing the non-modifying operation
+using the modifying operation.
+"""
+function similar_result(M::Manifold, VS::VectorSpaceType, f, x...)
+    T = similar_result_type(M, VS, f, x)
+    return similar(x[1], T)
+end
+
+
+"""
+    manifold_dimension(M::Manifold, VS::VectorSpaceType)
```

Comment:
Right, calling it `vector_space_dimension` will be more clear.

I was thinking about treating vector spaces as manifolds (there was that `VectorSpaceManifold` type) but it doesn't seem to be particularly useful. It was also a bit problematic to provide reliable implementations of `inner`, `exp`, `log`, etc for vector bundles, since either there would be two sets of methods for these vector spaces (one for `VectorSpaceManifold` and one general like the ones I currently have) or all methods would have to work on `VectorSpaceManifold`. In the first case extending these methods could lead to inconsistencies and in the second case new objects of that type would have to be needlessly constructed in many methods of `VectorBundle`. How would you solve this issue?

Reply:
To be honest, I haven't thought about it much, and I'm not certain I understand the problem or how you intend to use vector spaces. It seems to me that it's fine to by default not implement methods that aren't general for all vector spaces (e.g. `exp`). Although, correct me if I'm wrong, but vector spaces behave much like Euclidean space and therefore `exp` and `log` would simply be addition and subtraction by a vector, right?

Part of the problem might be that we currently abuse notation a bit. Each of our method signatures begins with a manifold, which is roughly interpreted as "related to this manifold". So even though the inner product is technically defined on the tangent bundle of a manifold, we define it on the manifold itself. Likewise, `exp` includes a point on the manifold and a point in its tangent space, which are different manifolds, but only the manifold with the point is provided. I think what we've done so far makes for a cleaner interface, and perhaps the approach you're using here is likewise preferred because it's cleaner and easier to remember.

Reply:
You are right, `exp`, `log`, etc. are simple to implement for vector spaces. The problem, though, is that someone might want to implement a vector space where `exp` is not performed by simple addition of arrays for whatever reason and he would expect vector bundles to use an appropriate method.

Your point on the abuse of notation is very appropriate here. The "related to this manifold" convention was mostly sufficient until now. We need a way to specify which vector spaces given vectors come from. I think the approach I'm using here will be insufficient once I get to operations on more than one vector space anyway, I'll look for a better one.

Reply:
I've introduced a new type to solve this problem `ManifoldVectorSpace`. I'm not really happy with this name as it's neither a manifold nor a vector space, but essentially corresponds to the set of all fibers of a vector bundle. This might solve our problems, I guess we'll see when I add tensor products and exterior products.

Reply:
I am not completely in all theory concerning fibres, but why then not just call it `VectorBundleFibers` then?

Reply:
Thanks, that's a much better name :+1: .